src/hooks/useTrades.ts

```diff
@@ -19,21 +25,69 @@ export function useTrades(): Trade[] {
+  const { userAddress, networkId } = useWalletConnection()
+
+  const intervalId = useRef<null | NodeJS.Timeout>(null)
 
   useEffect(() => {
-    // Check only up to `latest - CONFIRMATION_BLOCKS` to reduce chance of re-orgs
-    const toBlock = blockNumber && blockNumber - CONFIRMATION_BLOCKS
-
-    if (userAddress && networkId && toBlock && (!lastCheckedBlock || lastCheckedBlock < toBlock)) {
-      getTrades({
-        userAddress,
-        networkId,
-        fromBlock: lastCheckedBlock,
-        toBlock,
-      }).then(newTrades => dispatch(appendTrades(newTrades, toBlock)))
+    function updateTrades(): void {
+      console.log(`Updating trades. Last checked block [${lastCheckedBlock}]`)
+      if (userAddress && networkId) {
+        // Don't want to update on every block
+        // So instead, we get the latest block when the time comes
+        web3.eth.getBlock('latest').then(({ number: blockNumber }) => {
+          console.log(`Updating trades. Latest block [${blockNumber}]`)
+
+          getTrades({
+            userAddress,
+            networkId,
+            // fromBlock is inclusive. If set, add 1 to avoid duplicates, otherwise return undefined
+            fromBlock: !lastCheckedBlock ? lastCheckedBlock : lastCheckedBlock + 1,
+            toBlock: blockNumber,
+          }).then(newTrades =>
+            dispatch(newTrades.length > 0 ? appendTrades(newTrades, blockNumber) : updateLastCheckedBlock(blockNumber)),
+          )
+        })
+      }
+    }
+
+    function clear(): void {
+      if (intervalId.current) clearInterval(intervalId.current)
+      intervalId.current = null
     }
-  }, [userAddress, networkId, blockNumber, lastCheckedBlock, dispatch])
+
+    clear()
+
+    // Don't bother starting the timeouts unless we are properly connected
+    if (userAddress && networkId) {
+      // Let's try to be a bit smarter, shall we?
+      // - Gnosis Protocol exchange work in batches, that resolve at every 5min.
+      // - Solutions are accepted until min 4 of a batch
+      // Thus, there's really no point in checking for new trades on every block.
+      // Once per batch is enough!
+      // Also:
+      // - Due to how the network is, some blocks might disappear after awhile due to re-orgs
+      // - If we wait for awhile, miners will sort this out and we'll get only what's final.
+      // Thus, we check for new trades a few seconds after the 4min mark
+      const delay = getTimeRemainingInBatch() - 30
+
+      // Update now to not leave the interface empty
+      if (trades.length === 0 || (delay >= 0 && delay < 10)) updateTrades()
+
+      if (delay >= 0 && delay < 10) {
+        // If time left in batch within this window, start the interval
+        intervalId.current = setInterval(updateTrades, BATCH_TIME_IN_MS)
+      } else {
+        // Otherwise, set a timeout to, update, then set interval
+        intervalId.current = setTimeout(() => {
```

Comment:
`intervalId` can live inside `useEffect`, since it's only used inside and cleared between calls